### PR TITLE
apps/files: add scanSingle and scanBatch to api

### PR DESCRIPTION
**Pull request by Yetangitu**
This is an _experimental_ implementation of an api for file scanning.

Two commands are added to the apps/files api:

**GET /api/v1/scan/path/to/file/to/scan**
Initiate a scan for a single path
returns: number of folders/files scanned

**POST /api/v1/scan**
POST data: array 'path_list' of paths to scan
returns: number of folders/files scanned

In theory both commands could return error counts for invalid user or invalid path, but for some reason the exceptions (_ForbiddenException_ and _InvalidArgumentException_) which _$scanner_ is supposed to throw do not get thrown (or are caught in flight) when it gets fed with nonsensical paths.

The entry points are currently not marked with _@noadminrequired_ (ie. only admin users can initiate scans), this mostly to keep normal users from loading the system with unnecessary scans.

This code was ported from 8.2.2 to master without testing as I currently do not have a running instance of the latter. Try before you buy...

### CODE REVIEW

**Comment by PVince81** [comment]
@icewind1991

**Comment by Yetangitu** [comment]
Here's a modified version of the change generator script which can be used with this api:

https://gist.github.com/Yetangitu/ae1ddc79c831d9e3472b

**Comment by icewind1991** [comment]
Looks good :+1:

**Comment by Yetangitu** [comment]
@owncloud-bot this contribution is MIT licensed

File: apps/files/appinfo/routes.php [hunk 1]
```diff
@@ -48,6 +48,17 @@
 				'verb' => 'GET',
 				'requirements' => array('tagName' => '.+'),
 			),
+                        array(
+                                'name' => 'API#scanSingle',
+                                'url' => '/api/v1/scan/{path}',
+                                'verb' => 'GET',
+                                'requirements' => array('path' => '.+'),
+                        ),
+                        array(
```
Comment (PVince81) [inline apps/files/appinfo/routes.php]: Please use square brackets for arrays, for consistency

File: apps/files/controller/apicontroller.php [hunk 2]
```diff
@@ -45,6 +45,14 @@ class ApiController extends Controller {
 	private $tagService;
 	/** @var IPreview */
 	private $previewManager;
+        /** @var int */
+        protected $foldersCounter = 0;
```
Comment (PVince81) [inline apps/files/controller/apicontroller.php]: indent looks off, please use tabs

**Comment by PVince81** [comment]
One part I'm a bit worried about is about permissions. Who should be able to call these APIs ?

I guess only an admin should be able to. It also means the admin needs to know the full path of the users for which to do the rescan.

**Comment by DeepDiver1975** [comment]
public http apis need to be implemented using ocs api

**Comment by DeepDiver1975** [comment]
I'm closing this - as explained above public http apis have to be implemented via OCS or WebDAV.

Thanks a lot for your contribution! Please get in touch with us on IRC to discuss on how to make this happen. THX

**Comment by PVince81** [comment]
One idea for Webdav would simply to have a Webdav property on the file/folder inside the regular Webdav endpoint: "oc:dirty" or something to tell the scanner to rescan. (setting that property could reset the size to -1 internally...)

**Pull request closed without merging**